### ISSUE TITLE
option to pass in initial hessian

### ISSUE THREAD

**Issue by jordangarside** [jhrmnn/pyberny#13]
Could you possibly add an option to allow for an initial hessian to be passed in?

**Comment by azag0** [comment]
In Cartesian or the generated internal coordinates? The former would be relatively difficult, because the Cartesian Hessian would first need to be transformed to the internal one, which would require calculating second derivatives of the internal coordinates w.r.t. the Cartesian ones (the derivative of the B matrix). The latter is easy.

**Comment by jordangarside** [comment]
Yeah I was thinking Cartesian. With a program like [TensorMol](https://github.com/jparkhill/TensorMol) hessians can be computed semi-cheaply.

pyberny works well as-is with TensorMol, the approximate hessian just becomes too expensive for 100s of atoms.

Seems like too much work for now though.

**Comment by azag0** [comment]
Being able to calculate the Hessian relatively cheaply at each step opens a whole new world of possibilities regarding geometry optimization. For instance, the advantage of internal over Cartesian coordinates becomes smaller—the main goal of internal coordinates is to work in a space where the Hessian is mostly diagonal and hence easy to approximate. If you actually know the Hessian, I would assume that working directly in Cartesian space would be good enough, and possibly better than internal coordinates with approximate Hessian.

**Comment by azag0** [comment]
I'll leave this open for now.

**Comment by jordangarside** [comment]
Yeah I'll get back to you on this when all of it's easy to test/work with.

**Comment by jparkhill** [comment]
A trivial idea would be to compute an approximate partial hessian (which is also easy in TF)... Ie: only blocks of the hessian which are within a cutoff, and feed that into pyberny. In this way linear scaling could be achieved and thousands of atoms could be addressed.

Usually an optimizer is only weakly dependent on the quality of the hessian. But still if PyBerny didn't do a sparse matrix-vector product for the H**-1 . G operation, we wouldn't get the maximum possible scope.
